### DISCUSSION TITLE
Описание клапана mixproof с IO-Link интерфейсом подключения

### DISCUSSION THREAD

**Discussion by idzm** [savushkin-r-d/ptusa_main#339] [Q&A] [answered]
Необходимо решить задачу #338.

**Comment by idzm** [comment]
Вот предварительный вариант:
![Mixproof_state_diagram](https://user-images.githubusercontent.com/23375200/146887865-d56d1f43-e6b9-4bb5-bfdc-f2bf66c62b98.png)

@pupenasan - что скажешь?

**Reply by idzm** [reply]
Этот вариант не очень хороший, так как здесь смешаны два разных состояния - положение клапана и состояние аварии.

**Comment by dimabeel** [comment]
https://refactoring.guru/ru/design-patterns/state

**Comment by idzm** [comment]
Вот вариант после дискуссий:

![Mixproof_state_diagram](https://user-images.githubusercontent.com/23375200/147914207-99b04b4a-7fe9-4ee7-8f22-5218a5856ed6.png)

(НС - нижнее седло, ВС - верхнее седло, узел I/O - узел ввода/вывода)

**Comment by idzm** [comment]
Вот вариант от Виктора:

![Модель состояний МиксПруф](https://user-images.githubusercontent.com/23375200/147914839-e82b81a5-c1d6-46a3-a076-44c4e013e6a7.png)

**Reply by idzm** [reply]
@pupenasan - нужно твоё мнение.

**Reply by pupenasan** [reply]
> @pupenasan - нужно твоё мнение.

В твоём варианте смущает одно проемежуточное состояние "открываюсь" для 3-х вариантов: "открыт", "открыт НС", "открыт ВС". Я не знаю, как по факту идёт переход в позиции НС, ВС, т.е. с какого состояния?

**Reply by 1967vvm** [reply]
Правильно ли я понял: 
Из состояния "Закрыт" можно перейти в три других состояния, во время перемещения, которое может длиться 10-15 сек.,
отсутствуют сигналы с датчиков положения, отсюда и состояние "открываюсь". точно так же, реализован переход в состояние "закрыть".
Если же отсутствует связь с модулем ввода-вывода то переходим в состояние "неизвестно".
Куда переходим в случае неправильных положений датчиков положения, и других возможных аварий?

**Reply by idzm** [reply]
> > @pupenasan - нужно твоё мнение.
> 
> В твоём варианте смущает одно проемежуточное состояние "открываюсь" для 3-х вариантов: "открыт", "открыт НС", "открыт ВС". Я не знаю, как по факту идёт переход в позиции НС, ВС, т.е. с какого состояния?

Открытие седел идёт из позиции `Закрыт`.

**Comment by pupenasan** [comment]
В поищиции Виктора идут другие понятия. Я не знаю что они значат. 
Аварию я б тоже не смешивал, она может служить условием перехода в какую-то позицию, т.е. влиять на автомат состояния по позиции. Если ИМ это  учитывает (т.е. он с логикой правления), то он также управляет автоматом состояния. Тогда у него свой автомат, который учитывает состояния датчиков, команды управления, внутреннюю логику контрля алармов, и управляет.

**Reply by 1967vvm** [reply]
Так и есть, в голове управления ThinkTop® клапаном Альфа Лаваль, уже реализован автомат состояний. Я же попытался отобразить его в виде блок-схемы.

**Reply by idzm** [reply]
Если быть точнее, то клапан передаёт информацию о своём состоянии в следующем виде (из [документации](https://www.alfalaval.com/globalassets/documents/products/fluid-handling/automation/control-unit/thinktop/interface-description-thinktop-v70-.pdf?_ga=2.214681746.422795805.1641371176-1132064945.1641371175)):

![изображение](https://user-images.githubusercontent.com/23375200/148185607-7fb097bb-f301-46ec-834b-4ad0f460f0a9.png)

**Reply by idzm** [reply]
@pupenasan - нужно твоё мнение по последним комментариям.

**Reply by pupenasan** [reply]
кинь сюда мануал

**Reply by pupenasan** [reply]
нашел, ссылка уже тут есть

**Reply by pupenasan** [reply]
кинь ссылку на сам клапан, бо я так и не понял, про эти седла

**Reply by pupenasan** [reply]
это он? https://youtu.be/_lo-3MLRX0E

**Reply by idzm** [reply]
Вот он - https://www.alfalaval.ru/products/fluid-handling/automation/sensing-and-control/thinktop/

**Reply by pupenasan** [reply]
надо будет обговорить в голосовом режиме

**Comment by idzm** [comment, marked as answer]
После обсуждений вот какой вариант получается:

![Mixproof_state_diagram](https://user-images.githubusercontent.com/23375200/150156923-8b5fc020-a49d-4d7a-a559-5ee745fa2f97.png)

Файл [DIA](http://dia-installer.de/index.html.en):
[Mixproof_state_diagram_v2.zip](https://github.com/savushkin-r-d/ptusa_main/files/7898141/Mixproof_state_diagram_v2.zip)

**Reply by idzm** [reply]
@pupenasan, @1967vvm - что скажете?

**Reply by pupenasan** [reply]
как по мне - гуд

**Reply by idzm** [reply]
@1967vvm - твое мнение какое?

**Reply by pupenasan** [reply]
> @1967vvm - твое мнение какое?

А что поменялось? Я вроде сказал своё мнение.

**Reply by idzm** [reply]
@1967vvm - Это Виктор.

**Reply by 1967vvm** [reply]
1- Если говорим об "абстрактном" клапане то все хорошо, если же мы используем "голову " от ТетраПак, с имеющимся "автоматом состояний", то возможны накладки или не стыковки .
2- Нужно добавить "автомат состояний" при возникновении аварий.
